beetsplug/convert.py

```diff
@@ -461,8 +472,28 @@ def convert_func(self, lib, opts, args):
                 self.copy_album_art(album, dest, path_formats, pretend,
                                     link, hardlink)
 
-        self._parallel_convert(dest, opts.keep_new, path_formats, fmt,
-                               pretend, link, hardlink, threads, items)
+        self._parallel_convert(dest, opts.keep_new, path_formats, fmt, pretend,
+                               link, hardlink, threads, items)
+
+        if playlist:
+            # When playlist arg is passed create an m3u8 file in dest folder.
+            #
+            # The classic m3u format doesn't support special characters in
+            # media file paths, thus we use the m3u8 format which requires
+            # media file paths to be unicode. Additionally we use relative
+            # paths to ensure readability of the playlist on remote
+            # computers.
+            self._log.info("Creating playlist file: {0}",
+                           util.normpath(playlist))
+            items_paths = [
+                util.syspath(item.destination(
+                    basedir=dest, path_formats=path_formats, fragment=True
+                )) for item in items
+            ]
+            if not pretend:
+                m3ufile = M3UFile(playlist)
+                m3ufile.set_contents(items_paths)
+                m3ufile.write()
```

Comment:
@wisp3rwind may I ask for your advice on a specific thing here: I recently came across about a PR where you cleaned up a little and moved the wrapping through syspath to the very last thing happening _before_ a filesystem operation is done by beets: https://github.com/beetbox/beets/pull/4675/files#diff-1ab0ce830f9dd337e8c57f809d3488b4dfb78c12a61a2a526100038e05a7ca54R29

I think back when I wrote this feature / opened this PR I did not have a good understanding of what and why syspath() wrapping is required. Have a look at line 488 above.

In words: I'm generating a list of paths to media files that then will be saved into an m3u playlist file. I wrap every one of these paths into `syspath()`

I now think that this actually is not the purpose of syspath() since I do not actually do a filesystem operation from within beets. I should not use `syspath()` here. Is that correct?

Reply:
As an aside, I'm preparing to go through the codebase and convert everything to `pathlib` which will negate the need for syspath calls.

Reply:
Oh yeah pathlib! That is a cool package! I use it all the time in my projects! Great idea!

Reply:
Sounds right: Currently, beets' code represents every path internally as a `bytes` object. To keep things manageable, paths should be converted to this representation immediately upon ingestion from any source (CLI, input files, library APIs, ...), and converted to other representations as late as possible (usually just upon leaving beets' direct control, i.e. when writing them to a file or the DB, etc).

To be honest, I'm also not entirely sure without studying the code again what exactly is the right approach when receiving paths from, say, the commandline. Anyway, yes, `syspath` seems like the wrong function here, at least without further research on whether it does the right thing on each platform. The question to answer here would be how m3u files should normally be encoded in each case, and ensure that beets adheres to that.

As a matter of fact, does your code actually run on Linux?  At a first glance, I would expect that `M3UFile.write()` would crash when trying to join the paths (`bytes` as returned by `syspath`) with newlines.

Sorry for the late reply, I missed that you mentioned me here!

Reply:
Hi @wisp3rwind thanks a lot for the reply, late or not, it's helpful! :-)

One thing before I answer more: Is it possible that you are confusing the `syspath` function with `normpath` or `bytestring_path` (which is called by `normpath` for example.

Reason being: `syspath` doesn't do anything on Linux, it just returns as-is. As I understood `syspath`, it actually is ment for sanitizing Windows paths:

https://github.com/beetbox/beets/blob/2e18f84b380911c1d3ec2e5b76513b3924db2188/beets/util/__init__.py#L395-L426

https://github.com/beetbox/beets/blob/2e18f84b380911c1d3ec2e5b76513b3924db2188/beets/util/__init__.py#L138-L144

Reply:
> The question to answer here would be how m3u files should normally be encoded in each case, and ensure that beets adheres to that.
 
Exactly! I did my research back then when I started to code this around last summer. It turns out that there is no real standard of what m3u is. It is rather loose, therefore tools do what the dev's thought it should do. The best we can do is do a compromise to work for the most circumstances. The main points from my research are:

- The best summary / documentation seems to be this wikipdia page: https://en.wikipedia.org/wiki/M3U
  - Intersting points:
    - m3u contains **platform specific paths**
    - m3u**8** ensures paths are unicode! -> This is what I'm aiming for, this is how this feature is documented so far. So this is my actual review question: Am I aiming at this goal?
  - Even proprietory DJ Tools (which are one of my target tools as a DJ obviously) might not do a 100% the right thing, thus this feature needs some flexibility. (One tiny example: Native Instruments Traktor understands m3u8 formatted playlists (utf-8 paths) but it does not understand the file ending of m3u8)

Reply:
>

Reply:
> > As a matter of fact, does your code actually run on Linux? At a first glance, I would expect that `M3UFile.write()` would crash when trying to join the paths (`bytes` as returned by `syspath`) with newlines.
> 
> Code is tested on macOS Catalina and Manjaro Linux (my former dev machine and my current main dev machine ;-))
> 
> One thing before I answer more: Is it possible that you are confusing the `syspath` function with `normpath` or even `bytestring_path` (which is called by `normpath`)
> 
> Reason being: `syspath` doesn't do anything on Linux/Posix, it just returns as-is. As I understood `syspath`, it actually is ment for sanitizing Windows paths:


Ah, sorry, now I understand: Since internally we are having `bytes` paths,. `syspath`, even when not doing anything on Non-Windows systems, will return exactly that: The internatl representation of paths, we still have `bytes`

So to make my question more concrete now: My goal is to now make sure that all paths that goe in the m3u list should be converted to UTF-8. Which function from utils could I use?

`displayable_path()`? Sounds odd but actually it translates `bytes` to `unicode`.

Reply:
https://github.com/beetbox/beets/blob/2e18f84b380911c1d3ec2e5b76513b3924db2188/beets/util/__init__.py#L376-L392